### ISSUE TITLE
redux @@redux/INIT actions

### ISSUE THREAD

**Issue by SebastienDaniel** [omnidan/redux-ignore#4]
It's kind of counter-intuitive to have to manually add `@@redux/INIT` tests that pass when filtering actions, since this is an extension of redux default behaviour.

```js
filterActions(
        require("./session"),
        function(action) {
            return /@@redux|_SESSION$/g.test(action.type);
        }
    )
```

Shouldn't those actions be handled by default by this module?

**Comment by omnidan** [comment]
@SebastienDaniel Handling the `@@redux/INIT` action is [an anti-pattern and should not be used](https://github.com/reactjs/redux/issues/186).

**Comment by SebastienDaniel** [comment]
I guess I didn't express myself properly :o

I'm not saying we should handle or react to those actions in anyway whatsoever, I'm saying those actions should pass any controls set by redux-ignore. In other words, they shouldn't ever be ignored.

Or is that still an anti-pattern?

**Comment by omnidan** [comment]
@SebastienDaniel It's an anti-pattern to treat `@@redux/INIT` special in any way. The idea is that you shouldn't rely on that action to initialize your state properly. When the first real action comes in, the default value for the state will be used anyway.

**Comment by SebastienDaniel** [comment]
The code is working fine. Its mostly a boilerplate issue, so nothing critical. 

I felt that having to account for `@@redux/INIT` when using `redux-ignore` was an implementation detail of redux that should be handled by redux-ignore.

Here is what my root reducer looks like when using redux-ignore. 

```js
combineReducers({
    apis: ignore.ignoreActions(
        require("./apis"),
        function(action) {
            return /^ACTIVATE_|^FILTER_|^SORT_|^WAITING_FOR$|^TRIGGER_$|^ADD_$/g.test(action.type);
        }
    ),
    session: ignore.filterActions(
        require("./session"),
        function(action) {
            return /@@redux|_SESSION$/g.test(action.type);
        }
    ),
    components: require("./components"),
    shell: ignore.filterActions(
        require("./shell"),
        function(action) {
            return /@@redux|^ACTIVATE_|^TRIGGER_|^ADD_/g.test(action.type);
        }
    )
});
```

Notice that I NEED to account for `@@redux/INIT`, otherwise redux throws the error:
`Reducer "reducerName" returned undefined during initialization. If the state passed to the reducer is undefined, you must explicitly return the initial state. The initial state may not be undefined.` 
This is caused by `@@redux/INIT` actions being ignored by redux-ignore, preventing the store from returning it's initial state on redux initialization.

Maybe there's something I'm not understanding or a piece of information I'm missing. You seem to be saying that I'm not understanding the issue (*all the better if I can learn from this!*)

thx for your patience :)

**Comment by omnidan** [comment]
@SebastienDaniel ah, I understand your issue now, but unfortunately this isn't an easy fix.

I can think of two solutions to this:

1. Always let `@@redux/INIT` pass through - this could break in later versions of redux [if it is changed to a random string](https://github.com/reactjs/redux/issues/186#issue-91614454). (you are not meant to handle that action)
2. Let all `@@` actions pass through - this would solve the issue with 1. but might cause bugs when `@@` is used to namespace an action, e.g. `@@app/ACTION` - those actions would never be ignored

In both solutions, we are introducing lots of magic and causing more bugs than we are fixing. `@@redux/INIT` is [generally problematic with higher-order reducers](https://github.com/omnidan/redux-undo/issues/49).

I'll reopen this and add a `help wanted` tag in case someone comes up with a decent solution for this.

**Comment by liesislukas** [comment]
I've just added redux-ignore to my project and i instantly get mentioned error:

> VM35779:263 Uncaught Error: Reducer "user" returned undefined during initialization. If the state passed to the reducer is undefined, you must explicitly return the initial state. The initial state may not be undefined

**Comment by Mike-Dax** [comment]
I'm having the same issue as liesislukas, would that be related to this error?

**Comment by omnidan** [comment]
this should be fixed in [1.2.4](https://github.com/omnidan/redux-ignore/releases/tag/v1.2.4) (thanks @maxmechanic!) - please let me know if there are any more issues :grin:
